server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/lisk/sources/binance/BinanceRateSource.java

```diff
@@ -0,0 +1,107 @@
+ * the GPL2 ("Copyleft").
+ *
+ * Contact information
+ * -------------------
+ *
+ * GENERAL BYTES s.r.o.
+ * Web      :  http://www.generalbytes.com
+ *
+ ************************************************************************************/
+package com.generalbytes.batm.server.extensions.extra.lisk.sources.binance;
+
+import com.generalbytes.batm.server.extensions.Currencies;
+import com.generalbytes.batm.server.extensions.IRateSource;
+import com.generalbytes.batm.server.extensions.extra.lisk.wallets.liskbinancewallet.LskWallet;
+
+import si.mazi.rescu.RestProxyFactory;
+
+import java.math.BigDecimal;
+import java.util.*;
+
+import java.util.List;
+/**
+ * Created by kkyovsky on 11/29/17.
+ *
+ * Modified by kelvin on 8/20/2018
+ */
+
+public class BinanceRateSource implements IRateSource {
+    /**
+     * Expiry of cache in seconds
+     */
+    private BinanceAPI api;
+    private TetherPriceAPI apiUsdt;
+    private String preferredFiatCurrency = Currencies.USD;
+
+    public BinanceRateSource(String preferedFiatCurrency) {
+        api = RestProxyFactory.createProxy(BinanceAPI.class, "https://api.binance.com");
+        apiUsdt = RestProxyFactory.createProxy(TetherPriceAPI.class, "https://api.coingecko.com");
+
+        if (Currencies.USD.equalsIgnoreCase(preferedFiatCurrency)) {
+            this.preferredFiatCurrency = Currencies.USD;
+        }
+
+        if (Currencies.HKD.equalsIgnoreCase(preferedFiatCurrency)) {
+            this.preferredFiatCurrency = Currencies.HKD;
+        }
+    }
+
+    @Override
+    public Set<String> getCryptoCurrencies() {
+        Set<String> result = new HashSet<String>();
+        result.add(Currencies.BTC);
+        result.add(Currencies.BCH);
+        result.add(Currencies.LTC);
+        result.add(Currencies.ETH);
+        result.add(Currencies.DASH);
+        result.add(Currencies.XMR);
+        result.add(Currencies.LSK);
+
+        return result;
+    }
+
+    @Override
+    public Set<String> getFiatCurrencies() {
+        Set<String> result = new HashSet<String>();
+        result.add(Currencies.USD);
+        result.add(Currencies.HKD);
+        return result;
+    }
+
+    @Override
+    public String getPreferredFiatCurrency() {
+        return this.preferredFiatCurrency;
+    }
+
+    @Override
+    public BigDecimal getExchangeRateLast(String cryptoCurrency, String fiatCurrency) {
+        if (!getFiatCurrencies().contains(fiatCurrency)) {
+            return null;
+        }
+        final BinanceTickerData btcUsdt = api.getTicker("BTCUSDT");
+        final BinanceTickerData lskBtc = api.getTicker(cryptoCurrency + "BTC");
+        final List<Object> usdtFiat = apiUsdt.getTetherPrice(fiatCurrency.toLowerCase(), "tether");
+
+        if (usdtFiat != null && btcUsdt.getPrice()!=null && lskBtc.getPrice() !=null ) {
+            final Map<String, Object> usdtFiatJson = (Map<String, Object>) usdtFiat.get(0);
+            final Double lastUsdtFiat = (Double) usdtFiatJson.get("current_price");
```

Comment:
The rate still goes from JSON through double and thus loses precision. Don't query the result via the map-like interface, but define the correct return class (`TetherData`, I guess) with the correct field type of `BigDecimal` for `current_price`. This way the `BigDecimal` should be initialized directly from the JSON string and thus maintain the original precision.